### DISCUSSION TITLE
What is MISSING IN current TWEEN SYSTEM :)

### DISCUSSION THREAD

**Discussion by seppoday** [godotengine/godot-proposals#7901] [Engine Core]
So... Let's talk. What things do you miss in current Tween System? I think we can make pretty nice proposal after some feedback from you guys. I will try to update this post with every idea.

Ideas:
-
- Ability to set on/off for auto start and auto kill ([mikhaelmartin](https://github.com/godotengine/godot-proposals/discussions/7901#discussioncomment-7137168))
- Ability to connect tween via editor ([mikhaelmartin](https://github.com/godotengine/godot-proposals/discussions/7901#discussioncomment-7137168))

(Last update: 29.09.2023)

(Discussion created after that proposal:  Bring Tween Node back and Merge it with current Tween #7892 )

**Comment by mikhaelmartin** [comment]
1.  Ability to set on/off for auto start and auto kill.
- use case: ~~replay repetitive animation~~ restart animation without the need to create new tween. Currently we need to do validation check, kill, create_tween, then reappend tweeners. too much steps for a very common use case 
- use case: persistent signal connection. currently signal is broken every time tween is invalidated 

2. Ability to connect tween signal via editor (Impemented as Tween Node)
- use case: you need a decoupled connection between the tween signals with other nodes on the scene, without having to clutter scripts with a wrapper signal.

**Reply by SilencedPerson** [reply]
Doesn't `tween_callback()` cover first usecase? when the tween is set to loop?

**Reply by mikhaelmartin** [reply]
edited, i use wrong words

**Reply by mikhaelmartin** [reply]
this is what i meant

```GDScript
extends Node2D

var tween:Tween

func _on_damage():
	if tween:
		tween.kill()
		
	tween = create_tween()
	tween.tween_property($Sprite2D, "modulate", Color.RED, 0.1).from(Color.WHITE)
	tween.tween_property($Sprite2D, "modulate", Color.WHITE, 0.1)
```

**Reply by mikhaelmartin** [reply]
this is what i want 

```GDScript
extends Node2D

var tween:Tween

func _ready():
	tween = create_tween()
	tween.set_auto_start(false)
	tween.set_auto_kill(false)
	tween.tween_property($Sprite2D, "modulate", Color.RED, 0.1).from(Color.WHITE)
	tween.tween_property($Sprite2D, "modulate", Color.WHITE, 0.1)

func _on_damage():
	tween.stop()
	tween.play()
```

**Comment by timcode77** [comment]
Please improve the following warning. It is not possible to track down where the warning is emanating from:

> W 0:00:11:0502   start: Target object freed before starting, aborting Tweener.
>   <C++ Source>   scene/animation/tween.cpp:548 @ start()

**Reply by Calinou** [reply]
The warning is emitted here: https://github.com/godotengine/godot/blob/da945ce6266ce27ba63b6b08dc0eb2414594f7cb/scene/animation/tween.cpp#L548-L552

The issue is that the warning is only emitted if the object reference is invalid (i.e. the object is `null`), so there's unfortunately no additional information we can print about the object here. The only thing we could do is print the Tweener's `resource_name`, but the resource name is rarely defined by users (it's generally not used often). It's almost always a built-in resource, so printing its `resource_path` wouldn't be very useful either.

**Comment by Lortemanden** [comment]
In the editor under errors it will throw a error when a tween is started without any tweeners. All good and well but when you click on the error it just opens a few lines with some cpp nonsense that does nothing to help identify the problem. I think it would be great if it pointed to where the tween was created in the code instead. 

![image](https://github.com/godotengine/godot-proposals/assets/55434783/b0c58846-6a1e-44c0-a747-1a99cf7d7a4f)

**Reply by AThousandShips** [reply]
It's not simple to do that, and it's already tracked:
* https://github.com/godotengine/godot/issues/79055
